pkg/api/volumesnapshot/handler.go

```diff
@@ -52,18 +55,36 @@ func (h *ActionHandler) do(rw http.ResponseWriter, r *http.Request) error {
 		if input.Name == "" {
 			return apierror.NewAPIError(validation.InvalidBodyContent, "Parameter `name` is required")
 		}
-		return h.restore(r.Context(), snapshotNamespace, snapshotName, input.Name)
+		return h.restore(r.Context(), snapshotNamespace, snapshotName, input.Name, input.StorageClassName)
 	default:
 		return apierror.NewAPIError(validation.InvalidAction, "Unsupported action")
 	}
 }
 
-func (h *ActionHandler) restore(ctx context.Context, snapshotNamespace, snapshotName, newPVCName string) error {
+func (h *ActionHandler) restore(ctx context.Context, snapshotNamespace, snapshotName, newPVCName, storageClassName string) error {
 	volumeSnapshot, err := h.snapshotCache.Get(snapshotNamespace, snapshotName)
 	if err != nil {
 		return err
 	}
 
+	sourceStorageClassName := volumeSnapshot.Annotations[util.AnnotationStorageClassName]
+	sourceStorageProvisioner := volumeSnapshot.Annotations[util.AnnotationStorageProvisioner]
+	sourceImageId := volumeSnapshot.Annotations[util.AnnotationImageID]
+	// default to using source storageClass
+	if storageClassName == "" {
+		storageClassName = sourceStorageClassName
+	}
+	sc, err := h.storageClassCache.Get(storageClassName)
+	if err != nil {
+		return apierror.NewAPIError(validation.InvalidBodyContent, fmt.Sprintf("Can not restore volume snapshot with a storageClass that does not exist: %v", err))
+	}
+	if sourceStorageProvisioner != "" && sc.Provisioner != sourceStorageProvisioner {
+		return apierror.NewAPIError(validation.InvalidBodyContent, "Can not use different provisioner for restoring volume snapshot")
+	}
+	if sourceImageId != "" && storageClassName != sourceStorageClassName {
+		return apierror.NewAPIError(validation.InvalidBodyContent, "Can not use different storageClass for restoring image volume snapshot")
```

Comment:
I do not understand this error.
On the above comment, looks like we could restore with different storage classes?
If yes, why this checking would pop up the APIError?

Reply:
For image volumes, only support restore with same storage classes
For other volumes, support restore  with different storage classes

Reply:
OK, I see. Thanks for explaining it.